src/config/session.c

```diff
@@ -80,6 +83,74 @@ read_environment_file(const char *filename)
 	return true;
 }
 
+static char *
+valid_env_path(const char *prefix, const char *name)
+{
+	if (!prefix || string_null_or_empty(name)) {
+		return NULL;
+	}
+
+	/* Valid environment files always end in '.env' */
+	const char *suffix = strrchr(name, '.');
+	if (!suffix || strcmp(suffix, ".env") != 0) {
+		return NULL;
+	}
+
+	char *full_path = strdup_printf("%s/%s", prefix, name);
+	if (!full_path) {
+		return NULL;
+	}
+
+	/* Valid environment file must be a regular file */
+	struct stat statbuf;
+	if (stat(full_path, &statbuf) == 0 && S_ISREG(statbuf.st_mode)) {
+		return full_path;
+	}
+
+	free(full_path);
+	return NULL;
+}
+
+static bool
+read_environment_dir(const char *path_prefix)
+{
+	bool success = false;
+	char *path = strdup_printf("%s.d", path_prefix);
+
+	errno = 0;
+	DIR *envdir = opendir(path);
+
+	if (!envdir) {
+		if (errno != ENOENT) {
+			const char *err_msg = strerror(errno);
+			wlr_log(WLR_INFO,
+				"failed to read environment directory: %s",
+				err_msg ? err_msg : "reason unknown");
+		}
+
+		goto env_dir_cleanup;
+	}
+
+	struct dirent *dirent;
+	while ((dirent = readdir(envdir)) != NULL) {
+		char *env_file_path = valid_env_path(path, dirent->d_name);
```

Comment:
1. Consider `s/valid_env_path/path_validate_and_dup/` (just to help the reader understand the allocation). Just a thought though.
2. I'm happy to do `if (dirent->d_type != DT_REG) { continue; }`. It avoids the `stat()` later. Think we have to define `_DEFAULT_SOURCE`, but it's supported "everywhere". It is my preferred approach generally, but happy to go with your preference.
3. Consider supporting symlinks too. Either via `lstat()` or by adding `&& dirent->d_name != DT_LNK`

Reply:
I renamed the function `strdup_env_path_validate` to make the allocation more obvious and to be more descriptive about the particular paths being validated. I've also added some logic to check `dirent->d_type` to confirm that the path actually refers to a regular file (or a symlink to a regular file).

The `dirent->d_type` field will contain `DT_LNK` for symlinks whether or not they are valid, but it may also contain `DT_UNKNOWN` on some filesystems (XFS seems like one of them) that may opt not to tell you the type of the file when reading a directory. We can short-circuit all of the path validity checks if the directory entry is any other type. After the path validation, we can use `dirent->d_type == DT_REG` to short-circuit the `stat` call, but we sill have to fall back to using `stat` for links or unknown types. The `stat` call follows symbolic links, so valid symlinks will just work with this implementaiton.